src/Core/Services/MetadataProviders/CosmosSqlMetadataProvider.cs

```diff
@@ -187,26 +187,49 @@ private void ParseSchemaGraphQLFieldsForJoins()
         /// 4. Check if we get previous entity with join information, if yes append it to the current entity also
         /// 5. Recursively call this function, to process the schema
         /// </summary>
-        /// <param name="fields"></param>
-        /// <param name="schemaDocument"></param>
-        /// <param name="currentPath"></param>
-        /// <param name="previousEntity">indicates the parent entity for which we are processing the schema.</param>
-        private void ProcessSchema(IReadOnlyList<FieldDefinitionNode> fields,
+        /// <param name="fields">All the fields of an entity</param>
+        /// <param name="schemaDocument">Schema Documents, useful to get fields information of an entity</param>
+        /// <param name="currentPath">Generated path of an entity</param>
+        /// <param name="tableCounter">Counter used to generate table alias</param>
+        /// <param name="parentEntity">indicates the parent entity for which we are processing the schema.
+        /// It is useful to get the JOIN statement information and create further new statements</param>
+        /// <param name="visitedEntity"> Keeps a track of the path in an entity, to detect circular reference</param>
+        /// <remarks>It detects the circular reference in the schema while processing the schema and throws <seealso cref="DataApiBuilderException"/> </remarks>
+        private void ProcessSchema(
+            IReadOnlyList<FieldDefinitionNode> fields,
             Dictionary<string, ObjectTypeDefinitionNode> schemaDocument,
             string currentPath,
             IncrementingInteger tableCounter,
-            EntityDbPolicyCosmosModel? previousEntity = null)
+            EntityDbPolicyCosmosModel? parentEntity = null,
+            HashSet<string>? visitedEntity = null)
         {
             // Traverse the fields and add them to the path
             foreach (FieldDefinitionNode field in fields)
             {
+                // Create a tracker to keep track of visited entities to detect circular references
+                HashSet<string> trackerForField = new();
+                if (visitedEntity is not null)
+                {
+                    trackerForField = visitedEntity;
+                }
+
                 string entityType = field.Type.NamedType().Name.Value;
+
                 // If the entity is not in the runtime config, skip it
                 if (!_runtimeConfig.Entities.ContainsKey(entityType))
                 {
                     continue;
```

Comment:
if the entity is not present in the runtime config, the check for circular reference below will be skipped. Is that the intention? Will we allow circular references if the entity is not present in the config?

Reply:
It would fail during operation when customer try to use that entity. (existing behavior)

Reply:
it may not fail during operation if the entity - which has the circular reference but not in runtime Config is present as a child of another entity - present in the config.

Reply:
as discussed, used buildin function to check the type.